testing/cuda_stream.cu

```diff
@@ -0,0 +1,72 @@
+/*
+ *  Copyright 2023 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <nvbench/cuda_stream.cuh>
+#include <nvbench/device_manager.cuh>
+#include <nvbench/types.cuh>
+
+#include "test_asserts.cuh"
+
+#include <fmt/format.h>
+
+namespace
+{
+/**
+ * @brief Queries and returns the device id that the given \p cuda_stream is associated with
+ *
+ * @param cuda_stream The stream to get the device id for
+ * @return The device id that \p cuda_stream is associated with
+ */
+int get_device_of_stream(cudaStream_t cuda_stream)
+{
+  CUcontext ctx;
+  NVBENCH_DRIVER_API_CALL(cuStreamGetCtx(CUstream{cuda_stream}, &ctx));
+  NVBENCH_DRIVER_API_CALL(cuCtxPushCurrent(ctx));
+  CUdevice device_id{};
+  NVBENCH_DRIVER_API_CALL(cuCtxGetDevice(&device_id));
+  NVBENCH_DRIVER_API_CALL(cuCtxPopCurrent(&ctx));
+  return static_cast<int>(device_id);
```

Comment:
Are we ok with using the CUDA Driver API within our tests? I see other CUDA Driver API calls being guarded by the `NVBENCH_HAS_CUPTI` macro. 

I would like to verify that the underlying `cudaStream_t` is in fact associated with the correct cuda `device_id`. The only alternative I see is to launch a kernel on a given stream on a given device and check if it would return an error. I'm not sure how reliable that would be.

Reply:
We can absolutely do this in the tests -- does it work without CUPTI?

FYI, I've been asked to replace all of our CUDA API calls with driver API eventually, but this hasn't happened yet.

Reply:
I just checked and without CUPTI it will complain about missing driver API symbols:
```
/usr/bin/ld: testing/CMakeFiles/nvbench.test.cuda_stream.dir/cuda_stream.cu.o: in function `(anonymous namespace)::get_device_of_stream(CUstream_st*)':
tmpxft_00000122_00000000-6_cuda_stream.compute_86.cudafe1.cpp:(.text+0x3e): undefined reference to `cuStreamGetCtx'
```

I've now wrapped the test using the `NVBENCH_HAS_CUPTI` macro.